src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs

```diff
@@ -145,38 +142,54 @@ public async ValueTask CreateListenerAsync()
             throw new InvalidOperationException($"The listener needs to be initialized by calling {nameof(CreateListenerAsync)}.");
         }
 
-        try
+        while (!cancellationToken.IsCancellationRequested)
         {
-            var quicConnection = await _listener.AcceptConnectionAsync(cancellationToken);
+            try
+            {
+                var quicConnection = await _listener.AcceptConnectionAsync(cancellationToken);
+
+                if (!_pendingConnections.TryGetValue(quicConnection, out var connectionContext))
+                {
+                    throw new InvalidOperationException("Couldn't find ConnectionContext for QuicConnection.");
+                }
+                else
+                {
+                    _pendingConnections.Remove(quicConnection);
+                }
+
+                // Verify the connection context was created and set correctly.
+                Debug.Assert(connectionContext != null);
+                Debug.Assert(connectionContext.GetInnerConnection() == quicConnection);
+
+                QuicLog.AcceptedConnection(_log, connectionContext);
 
-            if (!_pendingConnections.TryGetValue(quicConnection, out var connectionContext))
+                return connectionContext;
+            }
+            catch (QuicException ex) when (ex.QuicError == QuicError.OperationAborted)
             {
-                throw new InvalidOperationException("Couldn't find ConnectionContext for QuicConnection.");
+                // OperationAborted is reported when an accept is in-progress and the listener is unbind/disposed.
+                QuicLog.ConnectionListenerAborted(_log, ex);
+                return null;
             }
-            else
+            catch (ObjectDisposedException ex)
             {
-                _pendingConnections.Remove(quicConnection);
+                // ObjectDisposedException is reported when an accept is started after the listener is unbind/disposed.
+                QuicLog.ConnectionListenerAborted(_log, ex);
+                return null;
+            }
+            catch (Exception ex)
+            {
+                // If the client rejects the connection because of an invalid cert then AcceptConnectionAsync throws.
+                // An error thrown inside ConnectionOptionsCallback can also throw from AcceptConnectionAsync.
+                // These are recoverable errors and we don't want to stop accepting connections.
+                QuicLog.ConnectionListenerAcceptConnectionFailed(_log, ex);
```

Comment:
> An error thrown inside ConnectionOptionsCallback can also throw from AcceptConnectionAsync.

Do exceptions originating from the  TlsHandshakeCallbackOptions.OnConnection 
get logged at a debug level when we're using the HttpsConnectionMiddleware?

Won't most exception types bubble out through AuthenticateAsServerAsync and get logged as an error by KestrelConnectionOfT? Ideally errors thrown by user code like the OnConnection callback should be logged at a higher level.

This is a fairly low-level API though, and it's best to get the fix in quickly, so I'll approve it either way.